File: SWEagent/src/environment/repo.ts
/**
 * Repository management
 * Converted from sweagent/environment/repo.py
 */

import { z } from 'zod';
import * as path from 'path';
import * as fs from 'fs';
import { execSync } from 'child_process';

import { parseGhRepoUrl } from '../utils/github';
import { getLogger } from '../utils/log';
import { AbstractDeployment } from './deployment';
import { Command, UploadRequest } from './runtime';

const logger = getLogger('repo');

/**
 * Repository protocol interface
 */
export interface Repo {
  baseCommit: string;
  repoName: string;
  copy(deployment: AbstractDeployment): Promise<void>;
  getResetCommands(): string[];
}

/**
 * Get git reset commands
 */
function getGitResetCommands(baseCommit: string): string[] {
  return [
    'git fetch',
    'git status',
    'git restore .',
    'git reset --hard',
    `git checkout ${baseCommit}`,
    'git clean -fdq',
  ];
}

/**
 * Pre-existing repository configuration
 */
export const PreExistingRepoConfigSchema = z.object({
  repoName: z.string().describe('The repo name (must be at root of deployment)'),
  baseCommit: z.string().default('HEAD'),
  type: z.literal('preexisting'),
  reset: z.boolean().default(true),
});

export type PreExistingRepoConfig = z.infer<typeof PreExistingRepoConfigSchema>;

export class PreExistingRepo implements Repo {
  repoName: string;
  baseCommit: string;
  reset: boolean;

  constructor(config: PreExistingRepoConfig) {
    this.repoName = config.repoName;
    this.baseCommit = config.baseCommit;
    this.reset = config.reset;
  }

  async copy(_deployment: AbstractDeployment): Promise<void> {
    logger.info(`Using pre-existing repository ${this.repoName}`);
  }

  getResetCommands(): string[] {
    if (!this.reset) {
      return [];
    }
    return getGitResetCommands(this.baseCommit);
  }
}

/**
 * Local repository configuration
 */
export const LocalRepoConfigSchema = z.object({
  path: z.string().transform((p) => path.resolve(p)),
  baseCommit: z.string().default('HEAD'),
  type: z.literal('local'),
});

export type LocalRepoConfig = z.infer<typeof LocalRepoConfigSchema>;

export class LocalRepo implements Repo {
  path: string;
  baseCommit: string;
  repoName: string;

  constructor(config: LocalRepoConfig) {
    this.path = config.path;
    this.baseCommit = config.baseCommit;
    this.repoName = path.basename(this.path).replace(' ', '-').replace("'", ''); // Match Python sanitization
  }

  private checkValidRepo(): void {
    if (!fs.existsSync(this.path)) {
      throw new Error(`Could not find git repository at path=${this.path}`);
    }

    const gitPath = path.join(this.path, '.git');
    if (!fs.existsSync(gitPath)) {
      throw new Error(`${this.path} is not a git repository`);
    }

    // Check if repo is dirty (has uncommitted changes)
    // Note: This is a simplified check - Python uses GitPython for this
    const status = execSync('git status --porcelain', { cwd: this.path }).toString();
    if (status.trim() && !process.env.PYTEST_CURRENT_TEST) {
      throw new Error(`Local git repository ${this.path} is dirty. Please commit or stash changes.`);
    }
  }

  async copy(deployment: AbstractDeployment): Promise<void> {
    this.checkValidRepo(); // Check valid repo first

    logger.info(`Copying local repository from ${this.path}`);

    // Upload repository to deployment
    await deployment.runtime.upload({
      sourcePath: this.path,
      targetPath: `/${this.repoName}`,
    } as UploadRequest);

    // Change permissions
    const result = await deployment.runtime.execute({
      command: `chown -R root:root ${this.repoName}`,
      shell: true,
    } as Command);

    if (result.exitCode !== 0) {
      const msg = `Failed to change permissions on copied repository (exit code: ${result.exitCode}, stdout: ${result.stdout}, stderr: ${result.stderr})`;
      throw new Error(msg);
    }
  }

  getResetCommands(): string[] {
    return getGitResetCommands(this.baseCommit);
  }
}

/**
 * GitHub repository configuration
 */
export const GithubRepoConfigSchema = z.object({
  githubUrl: z.string(),
  baseCommit: z.string().default('HEAD'),
  cloneTimeout: z.number().default(500),
  type: z.literal('github'),
});

export type GithubRepoConfig = z.infer<typeof GithubRepoConfigSchema>;

export class GithubRepo implements Repo {
  githubUrl: string;
  baseCommit: string;
  cloneTimeout: number;
  repoName: string;
  constructor(config: GithubRepoConfig) {
    // Handle short form github URLs (org/repo)
    if (config.githubUrl.split('/').length === 2 && !config.githubUrl.includes('://')) {
      this.githubUrl = `https://github.com/${config.githubUrl}`;
    } else {
      this.githubUrl = config.githubUrl;
    }

    this.baseCommit = config.baseCommit;
    this.cloneTimeout = config.cloneTimeout;

    const parsed = parseGhRepoUrl(this.githubUrl);
    this.repoName = `${parsed.owner}__${parsed.repo}`;
  }

  private getUrlWithToken(token: string): string {
    if (!token) {
      return this.githubUrl;
    }

    // Check if @ already in URL
    if (this.githubUrl.includes('@')) {
      logger.warn('Cannot prepend token to URL. "@" found in URL');
      return this.githubUrl;
    }

    // Insert token into URL for authentication
    const urlParts = this.githubUrl.split('://');
    if (urlParts.length === 2) {
      return `https://${token}@${urlParts[1]}`;
    }

    return this.githubUrl;
  }

  async copy(deployment: AbstractDeployment): Promise<void> {
    logger.info(`Cloning GitHub repository ${this.githubUrl}`);

    const token = process.env.GITHUB_TOKEN || '';
    const url = this.getUrlWithToken(token);

    // Execute git commands directly in the deployment (matches Python implementation)
    const commands = [
      `mkdir /${this.repoName}`,
      `cd /${this.repoName}`,
      'git init',
      `git remote add origin ${url}`,
      `git fetch --depth 1 origin ${this.baseCommit}`,
      'git checkout FETCH_HEAD',
      'cd ..',
    ];

    await deployment.runtime.execute({
      command: commands.join(' && '),
      timeout: this.cloneTimeout,
      shell: true,
      check: true,
    } as Command);
  }

  getResetCommands(): string[] {
    return getGitResetCommands(this.baseCommit);
  }
}

/**
 * Union type for all repo configurations
 */
export const RepoConfigSchema = z.discriminatedUnion('type', [
  PreExistingRepoConfigSchema,
  LocalRepoConfigSchema,
  GithubRepoConfigSchema,
]);

export type RepoConfig = z.infer<typeof RepoConfigSchema>;

/**
 * Factory function to create repo from simplified input
 */
export function repoFromSimplifiedInput(
  input: string,
  baseCommit: string = 'HEAD',
  type: 'local' | 'github' | 'preexisting' | 'auto' = 'auto',
): Repo {
  if (type === 'auto') {
    // Auto-detect type
    if (input.includes('github.com')) {
      type = 'github';
    } else if (fs.existsSync(input)) {
      type = 'local';
    } else {
      type = 'preexisting';
    }
  }

  switch (type) {
    case 'github':
      return new GithubRepo({ githubUrl: input, baseCommit, type: 'github', cloneTimeout: 500 });
    case 'local':
      return new LocalRepo({ path: input, baseCommit, type: 'local' });
    case 'preexisting':
      return new PreExistingRepo({ repoName: input, baseCommit, type: 'preexisting', reset: true });
    default:
      throw new Error(`Unknown repo type: ${type}`);
  }
}

File: cpp_merge/packages/integration/SWEagent/src/environment/repo.cpp
#include "/home/runner/work/elizaos-cpp/elizaos-cpp/SWEagent/src/environment/repo.h"

array<string> getGitResetCommands(string baseCommit)
{
    return array<string>{ std:("git fetch"), std:("git status"), std:("git restore ."), std:("git reset --hard"), std:("git checkout ") + baseCommit + string_empty, std:("git clean -fdq") };
};


PreExistingRepo::PreExistingRepo(PreExistingRepoConfig config) {
    this->repoName = config->repoName;
    this->baseCommit = config->baseCommit;
    this->reset = config->reset;
}

std::shared_ptr<Promise<void>> PreExistingRepo::copy(std::shared_ptr<AbstractDeployment> _deployment)
{
    logger->info(std:("Using pre-existing repository ") + this->repoName + string_empty);
    return std::shared_ptr<Promise<void>>();
}

array<string> PreExistingRepo::getResetCommands()
{
    if (!this->reset) {
        return array<any>();
    }
    return getGitResetCommands(this->baseCommit);
}

LocalRepo::LocalRepo(LocalRepoConfig config) {
    this->path = config->path;
    this->baseCommit = config->baseCommit;
    this->repoName = path->basename(this->path)->replace(std:(" "), std:("-"))->replace(std:("'"), string_empty);
}

void LocalRepo::checkValidRepo()
{
    if (!fs::existsSync(this->path)) {
        throw any(std::make_shared<Error>(std:("Could not find git repository at path=") + this->path + string_empty));
    }
    auto gitPath = path->join(this->path, std:(".git"));
    if (!fs::existsSync(gitPath)) {
        throw any(std::make_shared<Error>(string_empty + this->path + std:(" is not a git repository")));
    }
    auto status = execSync(std:("git status --porcelain"), object{
        object::pair{std:("cwd"), this->path}
    })->toString();
    if (AND((status->trim()), (!process->env->PYTEST_CURRENT_TEST))) {
        throw any(std::make_shared<Error>(std:("Local git repository ") + this->path + std:(" is dirty. Please commit or stash changes.")));
    }
}

std::shared_ptr<Promise<void>> LocalRepo::copy(std::shared_ptr<AbstractDeployment> deployment)
{
    this->checkValidRepo();
    logger->info(std:("Copying local repository from ") + this->path + string_empty);
    std::async([=]() { deployment->runtime->upload(as<std::shared_ptr<UploadRequest>>(object{
        object::pair{std:("sourcePath"), this->path}, 
        object::pair{std:("targetPath"), std:("/") + this->repoName + string_empty}
    })); });
    auto result = std::async([=]() { deployment->runtime->execute(as<std::shared_ptr<Command>>(object{
        object::pair{std:("command"), std:("chown -R root:root ") + this->repoName + string_empty}, 
        object::pair{std:("shell"), true}
    })); });
    if (result->exitCode != 0) {
        auto msg = std:("Failed to change permissions on copied repository (exit code: ") + result->exitCode + std:(", stdout: ") + result->stdout + std:(", stderr: ") + result->stderr + std:(")");
        throw any(std::make_shared<Error>(msg));
    }
    return std::shared_ptr<Promise<void>>();
}

array<string> LocalRepo::getResetCommands()
{
    return getGitResetCommands(this->baseCommit);
}

GithubRepo::GithubRepo(GithubRepoConfig config) {
    if (AND((config->githubUrl->split(std:("/"))->length == 2), (!config->githubUrl->includes(std:("://"))))) {
        this->githubUrl = std:("https://github.com/") + config->githubUrl + string_empty;
    } else {
        this->githubUrl = config->githubUrl;
    }
    this->baseCommit = config->baseCommit;
    this->cloneTimeout = config->cloneTimeout;
    auto parsed = parseGhRepoUrl(this->githubUrl);
    this->repoName = string_empty + parsed["owner"] + std:("__") + parsed["repo"] + string_empty;
}

string GithubRepo::getUrlWithToken(string token)
{
    if (!token) {
        return this->githubUrl;
    }
    if (this->githubUrl->includes(std:("@"))) {
        logger->warn(std:("Cannot prepend token to URL. "@" found in URL"));
        return this->githubUrl;
    }
    auto urlParts = this->githubUrl->split(std:("://"));
    if (urlParts->get_length() == 2) {
        return std:("https://") + token + std:("@") + const_(urlParts)[1] + string_empty;
    }
    return this->githubUrl;
}

std::shared_ptr<Promise<void>> GithubRepo::copy(std::shared_ptr<AbstractDeployment> deployment)
{
    logger->info(std:("Cloning GitHub repository ") + this->githubUrl + string_empty);
    auto token = OR((process->env->GITHUB_TOKEN), (string_empty));
    auto url = this->getUrlWithToken(token);
    auto commands = array<string>{ std:("mkdir /") + this->repoName + string_empty, std:("cd /") + this->repoName + string_empty, std:("git init"), std:("git remote add origin ") + url + string_empty, std:("git fetch --depth 1 origin ") + this->baseCommit + string_empty, std:("git checkout FETCH_HEAD"), std:("cd ..") };
    std::async([=]() { deployment->runtime->execute(as<std::shared_ptr<Command>>(object{
        object::pair{std:("command"), commands->join(std:(" && "))}, 
        object::pair{std:("timeout"), this->cloneTimeout}, 
        object::pair{std:("shell"), true}, 
        object::pair{std:("check"), true}
    })); });
    return std::shared_ptr<Promise<void>>();
}

array<string> GithubRepo::getResetCommands()
{
    return getGitResetCommands(this->baseCommit);
}

std::shared_ptr<AgentLogger> logger = getLogger(std:("repo"));
any PreExistingRepoConfigSchema = z->object(object{
    object::pair{std:("repoName"), z->string()->describe(std:("The repo name (must be at root of deployment)"))}, 
    object::pair{std:("baseCommit"), z->string()->default(std:("HEAD"))}, 
    object::pair{std:("type"), z->literal(std:("preexisting"))}, 
    object::pair{std:("reset"), z->boolean()->default(true)}
});
any LocalRepoConfigSchema = z->object(object{
    object::pair{std:("path"), z->string()->transform([=](auto p) mutable
    {
        return path->resolve(p);
    }
    )}, 
    object::pair{std:("baseCommit"), z->string()->default(std:("HEAD"))}, 
    object::pair{std:("type"), z->literal(std:("local"))}
});
any GithubRepoConfigSchema = z->object(object{
    object::pair{std:("githubUrl"), z->string()}, 
    object::pair{std:("baseCommit"), z->string()->default(std:("HEAD"))}, 
    object::pair{std:("cloneTimeout"), z->number()->default(500)}, 
    object::pair{std:("type"), z->literal(std:("github"))}
});
any RepoConfigSchema = z->discriminatedUnion(std:("type"), array<any>{ PreExistingRepoConfigSchema, LocalRepoConfigSchema, GithubRepoConfigSchema });

void Main(void)
{
}

MAIN
